File: ext-go-legacy/internal/flows/flows.go
package flows

import (
	"context"
	"errors"
	"sconcur/internal/dto"
	"sync"
)

type Flows struct {
	mutex sync.RWMutex
	flows map[string]*Flow

	// pool recycles Flow structs (and their activeTasks backing map) between a
	// DeleteFlow and the next InitFlow, so the common one-shot flow — the sync
	// path and every async coroutine's own flow — stops allocating a fresh Flow +
	// map on each call. Only the per-flow cancel context is still allocated (a
	// cancelled context cannot be reused). Safe because flow keys are never reused
	// (see Flow.reset). Per-Flows so it is dropped with the handler on Destroy.
	pool sync.Pool
}

func NewFlows() *Flows {
	return &Flows{
		flows: make(map[string]*Flow),
	}
}

func (f *Flows) InitFlow(handlerCtx context.Context, flowKey string, results chan *dto.Result) *Flow {
	f.mutex.Lock()
	defer f.mutex.Unlock()

	flow, ok := f.flows[flowKey]

	if !ok {
		flow = f.acquire(handlerCtx, flowKey, results)

		f.flows[flowKey] = flow
	}

	return flow
}

// acquire returns a recycled Flow re-armed for this key, or a new one when the
// pool is empty. Called under the Flows lock.
func (f *Flows) acquire(handlerCtx context.Context, flowKey string, results chan *dto.Result) *Flow {
	pooled := f.pool.Get()

	if pooled == nil {
		return NewFlow(handlerCtx, flowKey, results)
	}

	flow := pooled.(*Flow)

	flow.reset(handlerCtx, flowKey, results)

	return flow
}

func (f *Flows) GetFlow(flowKey string) (*Flow, error) {
	f.mutex.RLock()
	defer f.mutex.RUnlock()

	flow, ok := f.flows[flowKey]

	if ok {
		return flow, nil
	}

	return nil, errors.New("flow not found")
}

func (f *Flows) DeleteFlow(flowKey string) {
	f.mutex.Lock()
	defer f.mutex.Unlock()

	flow, ok := f.flows[flowKey]

	if !ok {
		return
	}

	flow.Cancel()
	delete(f.flows, flowKey)

	// Recycle the detached flow. Its key is retired for good (keys are never
	// reused), so any of its results still sitting in the buffered channel route
	// by a key GetFlow no longer knows and are dropped — they can never reach the
	// struct once it is re-armed for a new key.
	f.pool.Put(flow)
}

func (f *Flows) GetTasksCount() int {
	f.mutex.RLock()
	defer f.mutex.RUnlock()

	var count int

	for _, flow := range f.flows {
		count += flow.Count()
	}

	return count
}

func (f *Flows) Cancel() {
	f.mutex.Lock()
	defer f.mutex.Unlock()

	for _, flow := range f.flows {
		flow.Cancel()
	}

	f.flows = make(map[string]*Flow)
}

File: ext/src/flows/flows.rs
//! The registry of live flows, keyed by the flow key PHP sends.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use tokio_util::sync::CancellationToken;

use crate::flows::Flow;
use crate::tasks::ResultSink;

pub struct Flows {
    inner: Mutex<FlowsInner>,
}

struct FlowsInner {
    flows: HashMap<String, Arc<Flow>>,
    /// Recycles Flow structs (and their active-task set) between a delete_flow
    /// and the next init_flow, so the common one-shot flow — the sync path and
    /// every async coroutine's own flow — stops allocating a fresh Flow + set
    /// on each call. Safe because flow keys are never reused (see Flow::reset).
    /// A plain Vec, not a lock-free pool: the registry lock is already held on
    /// both paths, so one would buy nothing here.
    pool: Vec<Arc<Flow>>,
}

impl Flows {
    pub fn new() -> Self {
        Flows {
            inner: Mutex::new(FlowsInner {
                flows: HashMap::new(),
                pool: Vec::new(),
            }),
        }
    }

    pub fn init_flow(
        &self,
        handler_ctx: &CancellationToken,
        flow_key: &str,
        results: &ResultSink,
    ) -> Arc<Flow> {
        let mut inner = self.inner.lock().unwrap();

        if let Some(flow) = inner.flows.get(flow_key) {
            return flow.clone();
        }

        let flow = match inner.pool.pop() {
            Some(pooled) => {
                pooled.reset(handler_ctx, flow_key.to_string(), results.clone());

                pooled
            }
            None => Arc::new(Flow::new(handler_ctx, flow_key.to_string(), results.clone())),
        };

        inner.flows.insert(flow_key.to_string(), flow.clone());

        flow
    }

    pub fn get_flow(&self, flow_key: &str) -> Option<Arc<Flow>> {
        self.inner.lock().unwrap().flows.get(flow_key).cloned()
    }

    pub fn delete_flow(&self, flow_key: &str) {
        let mut inner = self.inner.lock().unwrap();

        let Some(flow) = inner.flows.remove(flow_key) else {
            return;
        };

        flow.cancel();

        // Recycle the detached flow. Its key is retired for good (keys are
        // never reused), so any of its results still sitting in the buffered
        // channel route by a key get_flow no longer knows and are dropped —
        // they can never reach the struct once it is re-armed for a new key.
        // Only the last holder may recycle it: a task still running keeps its
        // own Arc, and re-arming under it would swap the flow beneath a live
        // task.
        if Arc::strong_count(&flow) == 1 {
            inner.pool.push(flow);
        }
    }

    pub fn tasks_count(&self) -> i32 {
        self.inner
            .lock()
            .unwrap()
            .flows
            .values()
            .map(|flow| flow.count())
            .sum()
    }

    pub fn cancel(&self) {
        let mut inner = self.inner.lock().unwrap();

        for flow in inner.flows.values() {
            flow.cancel();
        }

        inner.flows.clear();
        inner.pool.clear();
    }
}
